packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -105,72 +93,196 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
+        uint256 balance = balances[_validator];
+        require(balance >= _amount, "ValidatorPool: insufficient balances");
+        balance = balance - _amount;
 
-        uint256 validatorDeposits = depositsOf[validator];
-        uint256 slashing = SLASHING_AMOUNT;
+        if (balance < MIN_BONDING_AMOUNT) {
+            uint256 lastValidatorIndex = validators.length - 1;
+            if (lastValidatorIndex > 0) {
+                uint256 validatorIndex = validatorIndexes[_validator];
+                address lastValidator = validators[lastValidatorIndex];
 
-        if (validatorDeposits < SLASHING_AMOUNT) {
-            slashing = validatorDeposits - SLASHING_AMOUNT;
+                validators[validatorIndex] = lastValidator;
+                validatorIndexes[lastValidator] = validatorIndex;
+            }
+            delete validatorIndexes[_validator];
+            validators.pop();
         }
 
-        depositsOf[validator] -= slashing;
-        depositsOf[challenger] += slashing;
+        balances[_validator] = balance;
+    }
 
-        emit Slashing(validator, challenger, validatorDeposits);
+    /**
+     * @notice Returns output submitter address.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     *
+     * @return Address of output submitter.
+     */
+    function getOutputSubmitter(uint256 _outputIndex) internal view returns (address) {
+        Types.CheckpointOutput memory output = L2_ORACLE.getL2Output(_outputIndex);
+        return output.submitter;
     }
 
     /**
-     * @notice Returns the deposit balance of validator.
+     * @notice Returns a bonding corresponding to the output index. Reverts if a bonding does not exist.
      *
-     * @param _validator The address of validator
+     * @param _outputIndex Index of L2 checkpoint output.
      *
-     * @return Deposit balance of validator.
+     * @return A bonding data.
      */
-    function balanceOf(address _validator) public view returns (uint256) {
-        return depositsOf[_validator];
+    function getBonding(uint256 _outputIndex) external view returns (Types.Bonding memory) {
+        Types.Bonding storage bonding = bonds[_outputIndex];
+        require(
+            bonding.amount > 0 && bonding.expiresAt > 0,
+            "ValidatorPool: bonding is not exists"
+        );
+        return bonding;
     }
 
-    function isValidator(address _validator) public view returns (bool) {
-        return validators[_validator] || _validator == TRUSTED_VALIDATOR;
+    /**
+     * @notice Returns the balances of given address.
+     *
+     * @param _addr Address of validator.
+     *
+     * @return Balance of given address.
+     */
+    function balanceOf(address _addr) external view returns (uint256) {
+        return balances[_addr];
+    }
+
+    function isValidator(address _addr) external view returns (bool) {
+        uint256 index = validatorIndexes[_addr];
+        return validators[index] == _addr;
     }
```

Comment:
don't need to check if the `_addr` is not zero address?

Reply:
It returns false always if the `_addr` is zero address.